File: occ2det.c
#define PRINTF_BINARY_PATTERN_INT8 "%c%c%c%c%c%c%c%c"
#define PRINTF_BYTE_TO_BINARY_INT8(i)    \
    (((i) & 0x80ll) ? '1' : '0'), \
    (((i) & 0x40ll) ? '1' : '0'), \
    (((i) & 0x20ll) ? '1' : '0'), \
    (((i) & 0x10ll) ? '1' : '0'), \
    (((i) & 0x08ll) ? '1' : '0'), \
    (((i) & 0x04ll) ? '1' : '0'), \
    (((i) & 0x02ll) ? '1' : '0'), \
    (((i) & 0x01ll) ? '1' : '0')

#define PRINTF_BINARY_PATTERN_INT16 \
    PRINTF_BINARY_PATTERN_INT8              PRINTF_BINARY_PATTERN_INT8
#define PRINTF_BYTE_TO_BINARY_INT16(i) \
    PRINTF_BYTE_TO_BINARY_INT8((i) >> 8),   PRINTF_BYTE_TO_BINARY_INT8(i)
#define PRINTF_BINARY_PATTERN_INT32 \
    PRINTF_BINARY_PATTERN_INT16             PRINTF_BINARY_PATTERN_INT16
#define PRINTF_BYTE_TO_BINARY_INT32(i) \
    PRINTF_BYTE_TO_BINARY_INT16((i) >> 16), PRINTF_BYTE_TO_BINARY_INT16(i)
#define PRINTF_BINARY_PATTERN_INT64    \
    PRINTF_BINARY_PATTERN_INT32             PRINTF_BINARY_PATTERN_INT32
#define PRINTF_BYTE_TO_BINARY_INT64(i) \
    PRINTF_BYTE_TO_BINARY_INT32((i) >> 32), PRINTF_BYTE_TO_BINARY_INT32(i)

#include <stdio.h>
#include <assert.h>
#include <stdbool.h>
#include <stdlib.h>

unsigned long n_combinaison(unsigned n, unsigned k) {
    //Try to avoid overflow by interleaving * and /
    //https://blog.plover.com/math/choose.html
    unsigned long r = 1;
    unsigned d;
    if (k > n) return 0;
    for (d=1; d <= k; d++, n--) {
            r = (r*n) / d;
     }
     return r;
}

void gen_permutation(unsigned const n_orbital, unsigned const n_int, unsigned const log_size_orbital_bucket, 
                    unsigned const * const occ, unsigned const n_alpha, 
                    unsigned long * const n_det, unsigned **l_det_alpha, unsigned **l_det_beta)
{
    // Will store the value 
    unsigned * const det_patern =  (unsigned*) malloc ( n_int * sizeof(unsigned));
    unsigned * const occ_if = (unsigned*) malloc ( n_orbital * sizeof(unsigned));
 
    unsigned i_int;
    for (i_int=0; i_int < n_int; i_int++){
        det_patern[i_int] = 0;
    }

    unsigned n_alpha_f = n_alpha;
    // Number of single occupied alpha orbital
    unsigned n_single_orbital = 0;

    unsigned idx;
    for (idx = 0 ; idx < n_orbital; idx++){
        switch(occ[idx]) {
            case 1 :
            {
                occ_if[n_single_orbital] = idx;
                n_single_orbital += 1;
                break;
            }
            case 2 :
            {
                unsigned const det_i =  (idx >> log_size_orbital_bucket);
                unsigned const det_p = idx - ( (det_i  -1) << log_size_orbital_bucket); 

                det_patern[det_i] |= 1 << det_p; // set bit to one
                n_alpha_f -= 1;
                break;
            } 
        } 
   }
    assert (n_single_orbital <= (sizeof(unsigned)*8));

    *n_det = n_combinaison(n_single_orbital, n_alpha_f);

    *l_det_beta =  (unsigned*) malloc ( (*n_det) * n_int * sizeof(unsigned));
    *l_det_alpha =  (unsigned*) malloc ( (*n_det) * n_int * sizeof(unsigned));

    // p represent the permutation of n_alpha_f orbital in the single occupated orbitals
    // Initialize it with the first valid occupations
    // who correspond to the bitmask with `n_alpha_f` right bits set
    // aka p = \sum_{n=0}^m 2^n =  2^{m+1} - 1 == (1 << m) - 1. 
    unsigned p = (1 << n_alpha_f ) - 1;

    unsigned long idx_d;
    unsigned i;

    // Set the value of the determinant
    for (idx_d= 0 ; idx_d < (*n_det); idx_d++){

        unsigned *alpha = *l_det_alpha+idx_d*n_int;
        unsigned *beta =  *l_det_beta+idx_d*n_int; 

        for (i_int=0; i_int < n_int; i_int++){
            beta[i_int] = det_patern[i_int];
            alpha[i_int] = det_patern[i_int];
        }
        
        for ( i = 0 ; i < n_single_orbital; i++){
            bool const o = ( p >> i ) & 1; //index i of permutation p
            unsigned * const det = o ? alpha: beta; //Determinant are zero-ed .Modify only if need to set the bit to 1

            idx = occ_if[i];
            unsigned const det_i =  (idx >> log_size_orbital_bucket);
            unsigned const det_p = idx - ( (det_i  -1) << log_size_orbital_bucket); 

            det[det_i]  |= 1 << det_p ; //Set the bit to one
        }

        // Compute the permutation
        // https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
        unsigned const t = p | (p - 1); 
        //p = (t + 1) | (((~t & -~t) - 1) >> (__builtin_ctz(p) + 1));
        p = (t + 1) | (((~t & (t+1)) - 1) >> (__builtin_ctz(p) + 1));
    }
}

int main(int argc, char **argv)
{
    const unsigned size_orbital_bucket = sizeof(unsigned)*8;
    const unsigned log_size_orbital_bucket = __builtin_ctz(size_orbital_bucket);
    const unsigned mode = atoi(argv[1]); 
    const unsigned n_orbital = atoi(argv[2]);;
    const unsigned n_alpha =  atoi(argv[3]);

    const unsigned n_int = n_orbital  / size_orbital_bucket  + 1 ;
    printf("n_int: %u\n", n_int);

    unsigned *occ = (unsigned*) malloc ( n_orbital * sizeof(unsigned));

    unsigned d, i = 0;
    for (i= 0 ; i < n_orbital; i++) {
        occ[i] = atoi(argv[i+4]);
    }
 
    unsigned long n_det;
    unsigned *l_det_alpha;
    unsigned *l_det_beta;

    gen_permutation(n_orbital, n_int, log_size_orbital_bucket, occ, n_alpha, &n_det, &l_det_alpha,  &l_det_beta);

    printf("n_det: %lu\n", n_det);
  
    // 0 == Production  mode ; 1 == Test mode 
    d = (mode) ? 0 : n_det; 

    for (d  ; d < n_det; d++){
        for (i =0 ; i < n_int; i++){
            printf(PRINTF_BINARY_PATTERN_INT32"\n", PRINTF_BYTE_TO_BINARY_INT32(l_det_alpha[d*n_int + i]));
        }
        for (i =0 ; i < n_int; i++){
             printf(PRINTF_BINARY_PATTERN_INT32"\n", PRINTF_BYTE_TO_BINARY_INT32(l_det_beta[d*n_int + i]));
        }
    }

    return 0;    
}

File: occ2det.f90
PURE FUNCTION n_combination(n,k) RESULT(r)
    IMPLICIT NONE

    INTEGER(4), INTENT(IN) :: n,k
    INTEGER(8)             :: r
    INTEGER(4)             :: d, n0
   
    IF (k > n) THEN
        r = 0
        RETURN 
    ELSE
        r = 1 
    ENDIF

    n0 = n
    DO d=1, k
        r = (r*n0) / d
        n0 = n0 - 1
    ENDDO
END FUNCTION n_combination

MODULE detMod  
        INTEGER(4), ALLOCATABLE :: l_det_alpha(:,:), l_det_beta(:,:)
        INTEGER(8) :: n_det      
END MODULE detMod


SUBROUTINE gen_permutation(n_orbital, n_int, log_size_orbital_bucket, occ, n_alpha)
      use detMod, ONLY : l_det_alpha, l_det_beta, n_det
      implicit none
      INTEGER(8) :: n_combination

      INTEGER(4), INTENT(IN) :: n_orbital, n_alpha, n_int,log_size_orbital_bucket
      INTEGER(1), DIMENSION(n_orbital), INTENT(IN) :: occ

      INTEGER(4) :: n_alpha_f, n_single_orbital
      INTEGER(4), DIMENSION(n_int) :: det_pattern

      INTEGER(4) :: idx, idx_d, i, det_i, det_p, p,t,tt
      LOGICAL :: o

      INTEGER(4), ALLOCATABLE :: occ_if(:)
      INTEGER(4) :: size_orbital_bucket

      size_orbital_bucket = ibset(0,log_size_orbital_bucket)
      det_pattern(:) = 0
      n_alpha_f = n_alpha
      n_single_orbital = 0
 
      ALLOCATE(occ_if(n_orbital))

      DO IDX = 0, n_orbital-1
        SELECT CASE (occ(idx+1))
            CASE (1)
                n_single_orbital = n_single_orbital + 1
                occ_if(n_single_orbital) = idx
            CASE (2)
                det_i = rshift(idx, log_size_orbital_bucket) + 1
                det_p = and(idx,size_orbital_bucket-1)

                det_pattern(det_i) = ibset(det_pattern(det_i),det_p)
                n_alpha_f = n_alpha_f - 1;
            END SELECT
      END DO

      IF (n_single_orbital > BIT_SIZE(p) * 8) THEN
            print*, 'To many permutation'
            STOP 1
      ENDIF

      n_det = n_combination(n_single_orbital, n_alpha_f);

      IF (.NOT.ALLOCATED(l_det_alpha)) THEN
        ALLOCATE(l_det_alpha(n_int,n_det))
        ALLOCATE(l_det_beta(n_int,n_det))
      ENDIF

      p = lshift(1, n_alpha_f) - 1
        
      DO idx_d = 1, n_det

        l_det_alpha(:,idx_d) = det_pattern(:)
        l_det_beta(:,idx_d) = det_pattern(:)
    
        DO i=1,n_single_orbital
            idx = occ_if(i) 
            det_i = rshift(idx, log_size_orbital_bucket) + 1
            det_p = and(idx,size_orbital_bucket-1)

            IF (btest(p,i-1)) THEN
                l_det_alpha(det_i,idx_d) = ibset(l_det_alpha(det_i,idx_d), det_p)
            ELSE
                l_det_beta(det_i,idx_d) = ibset(l_det_beta(det_i,idx_d), det_p)
            ENDIF
        ENDDO

        ! Compute the permutation
        ! https://graphics.stanford.edu/~seander/bithacks.html#NextBitPermutation
        t = or(p, p-1)
        tt = t+1
        p = or(tt, (rshift(and(not(t),tt) - 1, trailz(p) +1)))

       ENDDO

       DEALLOCATE(occ_if)


END SUBROUTINE


program hello
    use detMod, ONLY : l_det_alpha, l_det_beta, n_det

    implicit none

    INTEGER(1), ALLOCATABLE :: occ(:)
    INTEGER(4) :: n_orbital, n_alpha, n_int
    INTEGER(4), PARAMETER :: size_orbital_bucket = 32
    INTEGER(4), PARAMETER :: log_size_orbital_bucket = trailz(size_orbital_bucket)
   
    CHARACTER(LEN=64) :: buffer
    
    INTEGER :: i, d, d_init, mode

    REAL(8) :: t1, t0

    call get_command_argument(1, buffer)
    read(buffer, '(I4)') mode


    call get_command_argument(2, buffer)
    read(buffer, '(I4)') n_orbital

    call get_command_argument(3, buffer)
    read(buffer, '(I4)') n_alpha

    n_int =  n_orbital  / size_orbital_bucket  + 1 ;
    print*, "n_int:",  n_int

    ALLOCATE(occ(n_orbital))
    DO i= 1, n_orbital
        call get_command_argument(i+3, buffer)
        read(buffer, '(I1)') occ(i)
    ENDDO

    do i=1,100
      call gen_permutation(n_orbital, n_int, log_size_orbital_bucket, occ, n_alpha)
    enddo
    print*, 'n_det:', n_det

    IF (mode.EQ.0) THEN
        d_init = n_det+1
    ELSE
        d_init = 1
    ENDIF
 
    DO D=d_init,n_det
        DO I=1, n_int
             write(*, '(B32.32)') l_det_alpha(I,D)
        ENDDO

        DO I=1, n_int
             write(*, '(B32.32)') l_det_beta(I,D)
        ENDDO

    ENDDO 
end program hello
